File: src/test/java/org/cloudfoundry/credhub/auth/UserContextFactoryTest.java
package org.cloudfoundry.credhub.auth;

import org.cloudfoundry.credhub.CredentialManagerApp;
import org.cloudfoundry.credhub.util.DatabaseProfileResolver;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.security.oauth2.common.OAuth2AccessToken;
import org.springframework.security.oauth2.provider.OAuth2Authentication;
import org.springframework.security.oauth2.provider.OAuth2Request;
import org.springframework.security.oauth2.provider.authentication.OAuth2AuthenticationDetails;
import org.springframework.security.oauth2.provider.token.ResourceServerTokenServices;
import org.springframework.security.web.authentication.preauth.PreAuthenticatedAuthenticationToken;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.junit4.SpringRunner;

import java.security.Principal;
import java.security.cert.X509Certificate;
import java.time.Instant;
import java.util.Date;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

import static org.cloudfoundry.credhub.auth.UserContext.AUTH_METHOD_MUTUAL_TLS;
import static org.cloudfoundry.credhub.auth.UserContext.AUTH_METHOD_UAA;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.core.IsEqual.equalTo;
import static org.hamcrest.core.StringContains.containsString;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.when;

@RunWith(SpringRunner.class)
@ActiveProfiles(profiles = {"unit-test"}, resolver = DatabaseProfileResolver.class)
@SpringBootTest(classes = CredentialManagerApp.class)
public class UserContextFactoryTest {
  @MockBean
  private ResourceServerTokenServices tokenServicesMock;

  @Autowired
  private UserContextFactory subject;

  @Test
  public void fromAuthentication_readsFromOAuthDetails() throws Exception {
    OAuth2Authentication oauth2Authentication = setupOAuthMock("TEST_GRANT_TYPE");
    UserContext context = subject.createUserContext(oauth2Authentication);

    assertThat(context.getUserId(), equalTo("TEST_USER_ID"));
    assertThat(context.getUserName(), equalTo("TEST_USER_NAME"));
    assertThat(context.getIssuer(), equalTo("TEST_UAA_URL"));
    assertThat(context.getScope(), equalTo("scope1,scope2"));
    assertThat(context.getGrantType(), equalTo("TEST_GRANT_TYPE"));
    assertThat(context.getValidFrom(), equalTo(1413495264L));
    assertThat(context.getValidUntil(), equalTo(1413538464L));
    assertThat(context.getAuthMethod(), equalTo(AUTH_METHOD_UAA));
  }


  @Test
  public void fromAuthentication_handlesSuppliedToken() throws Exception {

    OAuth2Authentication oauth2Authentication = setupOAuthMock("TEST_GRANT_TYPE");

    UserContext context = subject.createUserContext(oauth2Authentication, "tokenValue");

    assertThat(context.getUserName(), equalTo("TEST_USER_NAME"));
    assertThat(context.getIssuer(), containsString("TEST_UAA_URL"));
    assertThat(context.getScope(), equalTo("scope1,scope2"));
    assertThat(context.getAuthMethod(), equalTo(AUTH_METHOD_UAA));
  }


  @Test
  public void fromAuthentication_handlesMtlsAuth() throws Exception {

    PreAuthenticatedAuthenticationToken mtlsAuth = setupMtlsMock();
    UserContext context = subject.createUserContext(mtlsAuth);

    assertThat(context.getUserName(), equalTo(null));
    assertThat(context.getUserId(), equalTo(null));
    assertThat(context.getIssuer(), equalTo(null));
    assertThat(context.getScope(), equalTo(null));
    assertThat(context.getValidFrom(), equalTo(1413495264L));
    assertThat(context.getValidUntil(), equalTo(1413538464L));
    assertThat(context.getClientId(), equalTo("CN=test_cn,OU=app:e054393e-c9c3-478b-9047-e6d05c307bf2"));
    assertThat(context.getAuthMethod(), equalTo(AUTH_METHOD_MUTUAL_TLS));
  }

  @Test
  public void getAclUser_fromOAuthPasswordGrant_returnsTheUserGuid() throws Exception {
    OAuth2Authentication oauth2Authentication = setupOAuthMock("password");
    UserContext context = subject.createUserContext(oauth2Authentication);

    assertThat(context.getActor(),
        equalTo("uaa-user:TEST_USER_ID"));
  }

  @Test
  public void getAclUser_fromOAuthClientGrant_returnsTheClientId() throws Exception {
    OAuth2Authentication oauth2Authentication = setupOAuthMock("client_credentials");
    UserContext context = subject.createUserContext(oauth2Authentication);

    assertThat(context.getActor(),
        equalTo("uaa-client:TEST_CLIENT_ID"));
  }

  @Test
  public void getAclUser_fromMtlsCertificate_returnsAppGuid() throws Exception {
    final PreAuthenticatedAuthenticationToken authenticationToken = setupMtlsMock();
    UserContext context = subject.createUserContext(authenticationToken);

    assertThat(context.getActor(),
        equalTo("mtls-app:e054393e-c9c3-478b-9047-e6d05c307bf2"));
  }


  private OAuth2Authentication setupOAuthMock(String grantType) {
    OAuth2Authentication authentication = mock(OAuth2Authentication.class);
    OAuth2Request oauth2Request = spy(new OAuth2Request(
        null,
        "TEST_CLIENT_ID",
        null,
        false,
        null,
        null,
        null,
        null,
        null));
    OAuth2AccessToken token = mock(OAuth2AccessToken.class);
    OAuth2AuthenticationDetails authDetails = mock(OAuth2AuthenticationDetails.class);

    Map<String, Object> additionalInformation = new HashMap<>();
    additionalInformation.put("user_id", "TEST_USER_ID");
    additionalInformation.put("user_name", "TEST_USER_NAME");
    additionalInformation.put("iss", "TEST_UAA_URL");
    additionalInformation.put("iat", 1413495264);

    Set<String> scopes = new HashSet<>();
    scopes.add("scope1");
    scopes.add("scope2");

    when(oauth2Request.getGrantType()).thenReturn(grantType);
    when(authentication.getDetails()).thenReturn(authDetails);
    when(authDetails.getTokenValue()).thenReturn("tokenValue");

    when(authentication.getOAuth2Request()).thenReturn(oauth2Request);
    when(token.getAdditionalInformation()).thenReturn(additionalInformation);
    when(token.getExpiration()).thenReturn(Date.from(Instant.ofEpochSecond(1413538464)));
    when(token.getScope()).thenReturn(scopes);

    when(tokenServicesMock.readAccessToken("tokenValue")).thenReturn(token);

    return authentication;
  }

  private PreAuthenticatedAuthenticationToken setupMtlsMock() {
    X509Certificate certificate = mock(X509Certificate.class);
    Principal principal = mock(Principal.class);
    PreAuthenticatedAuthenticationToken token = mock(PreAuthenticatedAuthenticationToken.class);

    when(certificate.getSubjectDN()).thenReturn(principal);
    when(principal.getName()).thenReturn("CN=test_cn,OU=app:e054393e-c9c3-478b-9047-e6d05c307bf2");

    when(certificate.getNotAfter()).thenReturn(Date.from(Instant.ofEpochSecond(1413538464L)));
    when(certificate.getNotBefore()).thenReturn(Date.from(Instant.ofEpochSecond(1413495264L)));
    when(token.getCredentials()).thenReturn(certificate);

    return token;
  }
}

File: components/auth/src/test/kotlin/org/cloudfoundry/credhub/auth/UserContextFactoryTest.kt
package org.cloudfoundry.credhub.auth

import org.cloudfoundry.credhub.CredhubTestApp
import org.cloudfoundry.credhub.auth.UserContext.Companion.AUTH_METHOD_MUTUAL_TLS
import org.cloudfoundry.credhub.auth.UserContext.Companion.AUTH_METHOD_UAA
import org.cloudfoundry.credhub.utils.DatabaseProfileResolver
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.core.IsEqual.equalTo
import org.hamcrest.core.StringContains.containsString
import org.junit.Test
import org.junit.jupiter.api.Assertions
import org.junit.runner.RunWith
import org.mockito.Mockito.mock
import org.mockito.Mockito.`when`
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.security.oauth2.jwt.Jwt
import org.springframework.security.oauth2.jwt.JwtClaimNames
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken
import org.springframework.security.web.authentication.preauth.PreAuthenticatedAuthenticationToken
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.context.junit4.SpringRunner
import java.security.cert.X509Certificate
import java.time.Instant
import java.util.Date
import javax.security.auth.x500.X500Principal

@RunWith(SpringRunner::class)
@ActiveProfiles(profiles = ["unit-test"], resolver = DatabaseProfileResolver::class)
@SpringBootTest(classes = [CredhubTestApp::class])
class UserContextFactoryTest {
    @Autowired
    private val subject: UserContextFactory? = null

    @Test
    @Throws(Exception::class)
    fun fromAuthentication_readsFromOAuthDetails() {
        val oauth2Authentication = setupOAuthMock("TEST_GRANT_TYPE")
        val context = subject!!.createUserContext(oauth2Authentication)

        assertThat<String>(context.userId, equalTo("TEST_USER_ID"))
        assertThat<String>(context.userName, equalTo("TEST_USER_NAME"))
        assertThat<String>(context.issuer, equalTo("TEST_UAA_URL"))
        assertThat<String>(context.scope, equalTo("scope1,scope2"))
        assertThat<String>(context.grantType, equalTo("TEST_GRANT_TYPE"))
        assertThat(context.validFrom, equalTo(1413495264L))
        assertThat(context.validUntil, equalTo(1413538464L))
        assertThat<String>(context.authMethod, equalTo(AUTH_METHOD_UAA))
    }

    @Test
    @Throws(Exception::class)
    fun fromAuthentication_handlesSuppliedToken() {
        val oauth2Authentication = setupOAuthMock("TEST_GRANT_TYPE")

        val context = subject!!.createUserContext(oauth2Authentication)

        assertThat<String>(context.userName, equalTo("TEST_USER_NAME"))
        assertThat<String>(context.issuer, containsString("TEST_UAA_URL"))
        assertThat<String>(context.scope, equalTo("scope1,scope2"))
        assertThat<String>(context.authMethod, equalTo(AUTH_METHOD_UAA))
    }

    @Test
    @Throws(Exception::class)
    fun fromAuthentication_handlesMtlsAuth() {
        val mtlsAuth = setupMtlsMock()
        val context = subject!!.createUserContext(mtlsAuth)

        assertThat<String>(
            context.userName,
            equalTo<String>(
                null,
            ),
        )
        assertThat<String>(context.userId, equalTo<String>(null))
        assertThat<String>(context.issuer, equalTo<String>(null))
        assertThat<String>(context.scope, equalTo<String>(null))
        assertThat(context.validFrom, equalTo(1413495264L))
        assertThat(context.validUntil, equalTo(1413538464L))
        assertThat<String>(context.clientId, equalTo("CN=test_cn,OU=app:e054393e-c9c3-478b-9047-e6d05c307bf2"))
        assertThat<String>(context.authMethod, equalTo(AUTH_METHOD_MUTUAL_TLS))
    }

    @Test
    @Throws(Exception::class)
    fun getAclUser_fromOAuthPasswordGrant_returnsTheUserGuid() {
        val oauth2Authentication = setupOAuthMock("password")
        val context = subject!!.createUserContext(oauth2Authentication)

        assertThat<String>(
            context.actor,
            equalTo("uaa-user:TEST_USER_ID"),
        )
    }

    @Test
    @Throws(Exception::class)
    fun getAclUser_fromOAuthClientGrant_returnsTheClientId() {
        val oauth2Authentication = setupOAuthMock("client_credentials")
        val context = subject!!.createUserContext(oauth2Authentication)

        assertThat<String>(
            context.actor,
            equalTo("uaa-client:TEST_CLIENT_ID"),
        )
    }

    @Test
    @Throws(Exception::class)
    fun getAclUser_fromMtlsCertificate_returnsAppGuid() {
        val authenticationToken = setupMtlsMock()
        val context = subject!!.createUserContext(authenticationToken)

        assertThat<String>(
            context.actor,
            equalTo("mtls-app:e054393e-c9c3-478b-9047-e6d05c307bf2"),
        )
    }

    @Test
    fun getAclUser_withInvalidGrantType_throwsException() {
        val oauth2Authentication = setupOAuthMock("client_credentials")
        val context = subject!!.createUserContext(oauth2Authentication)
        context.grantType = "bruce is crazy"

        Assertions.assertThrows(UserContext.UnsupportedGrantTypeException::class.java) {
            context.actor
        }
    }

    @Test
    fun getAclUser_withInvalidAuthMethod_throwsException() {
        val invalidAuthMethod = "not a valid auth method"
        val context =
            UserContext(
                "some-user-id",
                "some-user-name",
                "some-issuer",
                11223344,
                22334455,
                "some-client-id",
                "some-scope",
                "some-grant-type",
                invalidAuthMethod,
            )

        Assertions.assertThrows(UserContext.UnsupportedAuthMethodException::class.java) {
            context.actor
        }
    }

    private fun setupOAuthMock(grantType: String): JwtAuthenticationToken {
        val authentication = mock(JwtAuthenticationToken::class.java)
        val jwt = mock(Jwt::class.java)
        val scopeList = listOf("scope1", "scope2")

        val claims =
            mapOf(
                JwtClaimNames.SUB to "TEST_USER_ID",
                JwtClaimNames.ISS to "TEST_UAA_URL",
                "user_id" to "TEST_USER_ID",
                "user_name" to "TEST_USER_NAME",
                "grant_type" to grantType,
                "client_id" to "TEST_CLIENT_ID",
                JwtClaimNames.IAT to Instant.ofEpochSecond(1413495264),
                JwtClaimNames.EXP to Instant.ofEpochSecond(1413538464),
                "scope" to scopeList,
            )

        `when`(jwt.claims).thenReturn(claims)
        `when`(jwt.subject).thenReturn("TEST_USER_ID")
        `when`(authentication.principal).thenReturn(jwt)
        return authentication
    }

    private fun setupMtlsMock(): PreAuthenticatedAuthenticationToken {
        val certificate = mock(X509Certificate::class.java)
        val principal = mock(X500Principal::class.java)
        val token = mock(PreAuthenticatedAuthenticationToken::class.java)

        `when`(certificate.subjectX500Principal).thenReturn(principal)
        `when`(principal.name).thenReturn("CN=test_cn,OU=app:e054393e-c9c3-478b-9047-e6d05c307bf2")

        `when`(certificate.notAfter).thenReturn(Date.from(Instant.ofEpochSecond(1413538464L)))
        `when`(certificate.notBefore).thenReturn(Date.from(Instant.ofEpochSecond(1413495264L)))
        `when`(token.credentials).thenReturn(certificate)

        return token
    }
}
